doc/guide.md

```diff
@@ -0,0 +1,353 @@
+}
+#endif
+```
+
+The `[name].h` header file must include the syscalls header.
+
+### Defining a Callback for Asynchronous Operations
+
+For every type of callback the system call can generate, the library must have a
+function signature defined as a type. The format is:
+
+```c
+typedef void (*libtock_[name]_callback_[desc])(returncode_t, int);
+```
+
+### Functions for Operations
+
+Every operation the libtock library should expose must be defined. The return
+type should generally be `returncode_t`. As these are asynchronous operations,
+they should have the last argument be a callback function pointer.
+
+```c
+returncode_t libtock_[name]_[desc](<arguments>, libtock_[name]_callback_[desc] cb);
+```
+
+### Example
+
+
+For example, a library called "sensor" with a sensor read operation should look
+like this:
+
+Define a suitable callback for the library:
+
+```c
+typedef void (*libtock_sensor_callback_reading)(returncode_t, int);
+```
+
+Define a upcall function to be passed to the kernel:
+
+```c
+static void sensor_temp_upcall(int                          ret,
+                               int                          val,
+                               __attribute__ ((unused)) int unused0,
+                               void*                        opaque) {
+  libtock_sensor_callback_reading cb = (libtock_sensor_callback_reading) opaque;
+  cb(ret, val);
+}
+```
+
+Define an external function for applications to use to run the asynchronous
+operation:
+
+```c
+returncode_t libtock_sensor_read(libtock_sensor_callback_reading cb) {
+  returncode_t ret;
+
+  ret = libtock_sensor_set_upcall(sensor_temp_upcall, cb);
+  if (ret != RETURNCODE_SUCCESS) return ret;
+
+  ret = libtock_sensor_command_read();
+  return ret;
+}
+```
+
+## Synchronous APIs
+
+Most system call interfaces will want to provide a synchronous API as well.
+
+| Characteristic   | Value                              |
+|------------------|------------------------------------|
+| Location         | `libtock-sync/[category]`          |
+| Source File Name | `libtock-sync/[category]/[name].c` |
+| Header File Name | `libtock-sync/[category]/[name].h` |
+
+### Header Files
+
+The libtock-sync `[name].h` header file must look like:
+
+```c
+#pragma once
+
+#include <libtock/tock.h>
+#include <libtock/[category]/[name].h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Function signatures go here.
+
+#ifdef __cplusplus
+}
+#endif
+```
+
+### Synchronous APIs
+
+For our sensor example:
+
+Create a struct to be set by the asynchronous callback:
+
+```c
+struct data {
+  bool fired;
+  int val;
+  returncode_t result;
+};
+
+static struct data result = { .fired = false };
+```
+
+Define a static callback function to pass to the asynchronous implementation:
+
+```c
+static void sensor_cb(returncode_t ret, int val) {
+  result.val = val;
+  result.fired = true;
+  result.result = ret;
+}
+```
+
+Define the external function in the `libtocksync_` name space for the sync
```

Comment:
Why do you want every function to have `libtock_` prepended? This seems more verbose for no benefit. I understand the purpose of putting `sync`, but not `libtock`.

Reply:
C doesn't really give you a choice since it doesn't have a namespace concept. If your driver is implementing a timer, `libtock_timer` is a unique name but `timer` is not.

Reply:
If you are writing a "pure" libtock-c app, where you are only calling libtock functions, I agree the prefix is a bit redundant. But for more complicated apps that use multiple libraries, it becomes very convenient to have namespacing to make the code easier to read.

I think we should get to a place where someone is writing a c app with some other library (e.g. lora or graphics or ML) and then just happens to use libtock, maybe for sensor data.